design/continuous-delivery.md

```diff
@@ -0,0 +1,578 @@
+
+> **Alternative considered**: one way to implement environment-specific name-spacing would have been to export the bootstrapping resources through a CloudFormation Export and then reference them using Fn::ImportValue. This would have worked for templates, but means that we would need a way to resolve import values during publishing as well (and as a result also Fn::Join, etc).
+
+**User Interface**
+
+This is the proposed API for `cdk-bootstrap`:
+
+```
+$ cdk bootstrap --profile XXX [--yes] aws://account/region --trust-account ACCOUNT_ID
+WARNING: any principal from <ACCOUNT_ID> will have administrative access to the account <account>.
+Please confirm (Y/N):
+```
+
+We should consider allowing users to specify a profile map that will allow bootstrapping multiple environments at the same time, but this can easily be achieved through a shell script:
+
+```bash
+#!/bin/sh
+cdk bootstrap --yes --profile prod-us aws://111111111111/us-east-1 --trust-account $PIPELINE_ACCOUNT
+cdk bootstrap --yes --profile prod-eu aws://222222222222/eu-west-2 --trust-account $PIPELINE_ACCOUNT
+```
+
+## Mutation
+
+Deployment of complex cloud application often involves a business-specific process which includes rolling out the app throughout multiple deployment phases and environments. This means that there is strong coupling between the structure of the application and the structure of its deployment pipeline. To enable users to represent this relationship naturally in their code, the CDK should support defining the app's deployment infrastructure as part of the CDK app.
+
+Therefore, we recommend that all resources needed for the deployment pipeline are defined as part of the CDK app itself in one or more stacks. After synthesis, the cloud assembly will include a set of CloudFormation templates for the pipeline stack(s).
+
+We can't begin to deploy an app before we provision and update the required the deployment resources based on the structure of the app. This is the purpose of the "**mutation**" stage.
+
+The initial creation of the pipeline will be performed manually using `cdk deploy pipeline-main` (where `pipeline-main` is name of the main pipeline stack), but from that point forward, any changes to the pipeline will be done by pushing a commit into the repo, and letting the pipeline pick it up.
+
+For example, if we use CodePipeline for deploying an app to multiple environments,the deployment infrastructure will consist of a central pipeline stack, which contains the pipeline itself, it's artifacts bucket and other related resources such as CodeBuild projects. It will also require a stack in each region that includes a CodePipeline regional replication bucket (and key).
+
+In CodePipeline, we will implement this stage using a CodeBuild action which runs `cdk deploy "pipeline-*"`. This will deploy all stacks that begin with the `pipeline-` prefix. These stacks can be deployed to any bootstrapped environment since `cdk deploy` can assume the deployment role.
```

Comment:
This needs Administrator permissions for the project's role that executes the `cdk deploy`, right?

Reply:
Added a note about mitigating the security risk